### DISCUSSION TITLE
Why IntervalVar instance has no access to start, end and duration LinearExpr instances

### DISCUSSION THREAD

**Discussion by gregy4** [google/or-tools#2847] [CP-SAT questions] [answered]
Hello,
  in java is IntervalVar class that is used for modelling of intervals in constraints noOverlap, cumulative and many others.
  From user point of view it may be helpful to have access to start, end and duration LinearExpr instances used during creation of IntervalVar. But IntervalVar has access only to internal LinearExpressionProto.
  Another usecase is to use IntervalVar for access to solution values of start, end and duration expressions. So far it was possible via CpSolverSolutionCallback.solutionIntegerValue(intervalVar.getBuilder().getStart) but change from IntVar to LinearExpr changed it. Instead of builder.getStart should be called builder.getStartView but it returns LinearExprProto that cannot be used as an argument of CpSolverSolutionCallback.value method. In issue #2832 was added variant for LinearExpr and not LinearExprProto. I see two variants, IntervalVar should return LinearExpr for start, end and duration or value method should accept LinearExprProto.

  Is it possible to have better access to IntervalVar internals or is there some reason why it is not possible ?

Best regards,
Jan

**Comment by lperron** [comment, marked as answer]
Implemented on master.

**Reply by Mizux** [reply]
For the record:
.Net https://github.com/google/or-tools/commit/44e22e50c760da1b3b1d544c5d2f6e9f61f06899#diff-dbb0c1ce438dca3a0141d0f01cab70da7ee709c4bfb58e02aee4a0d1f791ccde
Java: https://github.com/google/or-tools/commit/c25036996f686bf7091ad8c4fed48e3d6d643dfa#diff-1adf39da728bf5ce19e59f976ba6abe213a2fe79e7baf01760663ad288ad1ad5
Python: https://github.com/google/or-tools/commit/67cc8ed591acba1f9dc563832939b703b5b2324f#diff-62619217b98d0ec1e7716b4da63196ad2005fa441023933bb07d3b54a94d09a3R839-R850
